File: C/14_Backtracking/BOJ_16938_캠프준비.c
#include <stdio.h>
#include <stdlib.h>
#define math_max(a, b) a > b ? a : b
#define math_min(a, b) a < b ? a : b
int n, l, r, x, cnt = 0, *arr;
int compare(const void* a, const void* b) {
	int o1 = *(int*)a;
	int o2 = *(int*)b;
	if (o1 > o2) return 1;
	else if (o1 < o2) return -1;
	else return 0;
}
void dfs(int idx, int dpth, int min, int max, int sum) {
	if (dpth >= 2) {
		if (l <= sum && sum <= r && max - min >= x) {
			cnt++;
		}
	}
	for (int i = idx; i < n; i++) {
		int tmp_min = math_min(min, arr[i]), tmp_max = math_max(max, arr[i]);
		dfs(i + 1, dpth + 1, tmp_min, tmp_max, sum + arr[i]);
	}
}
main() {
	scanf("%d %d %d %d", &n, &l, &r, &x);
	arr = (int*)malloc(sizeof(int) * n);
	for (int i = 0; i < n; i++) {
		scanf("%d", &arr[i]);
	}
	qsort(arr, n, sizeof(int), compare);
	dfs(0, 0, 1000000, 0, 0);
	printf("%d", cnt);
}

File: JAVA/14_Backtracking/BOJ_16938_캠프준비.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static int n, l, r, x, cnt = 0;
    static int[] arr;

    public static void dfs(int idx, int dpth, int min, int max, int sum) {
        if (dpth >= 2) {
            if (l <= sum && sum <= r && max - min >= x) {
                cnt++;
            }
        }
        for (int i = idx; i < n; i++) {
            dfs(i + 1, dpth + 1, Math.min(min, arr[i]), Math.max(max, arr[i]), sum + arr[i]);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        n = Integer.parseInt(st.nextToken());
        l = Integer.parseInt(st.nextToken());
        r = Integer.parseInt(st.nextToken());
        x = Integer.parseInt(st.nextToken());
        arr = new int[n];
        st = new StringTokenizer(br.readLine(), " ");
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(arr);
        dfs(0, 0, Integer.MAX_VALUE, 0, 0);
        System.out.println(cnt);
    }
}